content/security.md

```diff
@@ -453,6 +454,84 @@ It's best to handle the redirection from HTTP to HTTPS on the platform which han
 
 In the event that a platform does not offer the ability to configure this, the `force-ssl` package can be added to the project and Meteor will attempt to intelligently redirect based on the presence of the `x-forwarded-for` header.
 
+<h2 id="httpheaders">HTTP Headers</h2>
+
+HTTP headers can be used to improve the security of apps, although these are not a silver bullet, they will assist users in mitigating more common attacks.  
+
+<h4 id="helmet">Recommended: Helmet</h4>
+
+Although there are many great open source solutions for setting HTTP headers, Meteor recommends [Helmet](https://helmetjs.github.io/). Helmet is a collection of 12 smaller middleware functions that set HTTP headers.
+
+First, install helmet.
+
+```js
+  meteor npm install helmet --save
+```
+
+By default, Helmet can be used to set various HTTP headers (see link above). These are a good starting point for mitigating common attacks. To use the default headers, users should use the following code anywhere in their server side meteor startup code.     
+
+> Note: Meteor has not extensively tested each header for compatibility with Meteor. Only headers listed below have been tested.
+
+```js
+// With other import statements
+import helmet from "helmet";
+
+// Within server side Meter.startup()
+WebApp.connectHandlers.use(helmet())
+```
+
+At a minimum, Meteor recommends users to set the following headers. Note that code examples shown below are specific to Helmet.
+
+<h3 id="csp">Content Security Policy</h3>
+
+> Note: Content Security Policy is not configured using Helmet's default header configuration.
+
+From MDN, Content Security Policy (CSP) is an added layer of security that helps to detect and mitigate certain types of attacks, including Cross Site Scripting (XSS) and data injection attacks. These attacks are used for everything from data theft to site defacement or distribution of malware.
+
+It is recommended  that users use CSP to protect their apps from access by third parties. CSP assists to control how resources are loaded into your application.
+
+By default, Meteor recommends unsafe inline scripts and styles are allowed, since many apps typically use them for analytics, etc. Unsafe eval is disallowed, and the only allowable content source is same origin or data, except for connect which allows anything (since meteor apps make websocket connections to a lot of different origins). Browsers will also be told not to sniff content types away from declared content types.
+
+```js
+// With other import statements
+import helmet from "helmet";
+
+// Within server side Meter.startup()
+WebApp.connectHandlers.use(
+  helmet.contentSecurityPolicy({
+    directives: {
+      defaultSrc: ["'self'"],
+      scriptSrc: ["'self'", "'unsafe-inline'"],
+      connectSrc: ["*"],
+      imgSrc: ["'self'"],
+      styleSrc: ["'self'", "'unsafe-inline'"],
+    },
+    browserSniff: false
```

Comment:
@dhrubins If you have a moment, do you think you could share your (likely more experienced) CSP eyes 👀 with this ruleset recommendation?

Reply:
Yeah, I agree that the default policy there is a reasonable baseline, but unfortunately CSP can get complex very quickly, so perhaps there could also be some links to CSP tools to help folks through the tough parts. Here are a few that helped me: [Google's CSP Evaluator](https://csp-evaluator.withgoogle.com/), [Report-URI's CSP Builder](https://report-uri.com/home/generate), [CSP documentation from Mozilla](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src), [CSPValidator.org](https://cspvalidator.org). Plus, the awesome [SecurityHeaders.com](https://securityheaders.com/) since helmet also helps you set other headers aside from CSP.

One Meteor specific challenge we had @LegalRobot was with removing unsafe-inline, which seems like a reasonable first step from the baseline in this PR. We switched to a mixture of script hashes and origins in our script-src directive, but we quickly found that this causes the Meteor runtime config script to get blocked (I don't remember if it was the lack of unsafe-inline, or the CSP rule that once you use a hash that everything needs a hash). Anyway, it effectively blocks Meteor from loading anything to the client after that. The only way we've figured out around this is to take the hash of the runtime config upon startup (and we needed it in base64 so we used node's native crypto module rather than the Meteor sha package, as is standard in Meteor core):

```
import crypto from 'crypto'
import { Autoupdate } from 'meteor/autoupdate'

Meteor.startup(() => {
  const runtimeConfig = Object.assign(__meteor_runtime_config__, Autoupdate)
  runtimeConfig.accountsConfigCalled = true
  const runtimeConfigHash = crypto.createHash('sha256').update(`__meteor_runtime_config__ = JSON.parse(decodeURIComponent("${encodeURIComponent(JSON.stringify(runtimeConfig))}"))`).digest('base64')
  ...
})
```

Now our helmet config for the script-src directive includes 'sha256-${runtimeConfigHash}', which just feels kinda hack-y (but, hey, it works!). We also found that we had to run this in the startup callback, rather than during the build.

We also had to conditionally disable some directives when running locally. I'm documenting our whole helmet config line by line and will post it shortly to hopefully save some other folks some pain when implementing the same.

Reply:
Here's a gist with our settings and how we use helmet: https://gist.github.com/dhrubins/31248506873a006ba5e6b459a81855c2

Writeup coming soon.